### PR TITLE
PBKDF2 CMAC implementation

### LINKED ISSUE #7645: PBKDF2: add support for CMAC
In #7132 we introduced an implementation of PBKDF2 based on `psa_mac_xxx()` APIs, which should work for both HMAC and CMAC - however we only tested HMAC.

Add test cases using CMAC, and fix any issue that might be found in the implementation when CMAC is used.

Note: if this issue done after #7133, it should have test cases for both `output_bytes()` and `output_key()` with CMAC.

**Depends on:** #7132

### PR DESCRIPTION
**Pull request by silabs-Kusumit**
## Description

Add implementation and test cases for `PSA_ALG_PBKDF2_AES_CMAC_PRF_128`

Fixes #7645 

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required
- [ ] **tests** provided

### CODE REVIEW

**Comment by silabs-Kusumit** [comment]
For the accelerated cipher configuration, should `MBEDTLS_PSA_BUILTIN_ALG_CMAC` be defined even if `PSA_WANT_ALG_CMAC` is not defined, given that `PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128` is defined?

**Comment by gilles-peskine-arm** [comment]
I think we should automatically define `PSA_WANT_ALG_CMAC` and `PSA_WANT_KEY_TYPE_AES` if `PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128` is defined, since an implementation of PBKDF2-CMAC-AES has to include an implementation of CMAC and AES. Although I'm not sure about the key type, since `PSA_WANT_KEY_TYPE_AES` promises all key sizes but `PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128` only requires AES-128.

Typically the WANT system handles this “automatically”, because the algorithm would be something like `PSA_ALG_PBKDF2(mac_alg)`, but there's no way to do the parametrization on AES there since no AES key is externally visible. So we'd need a `PSA_ALG_PBKDF2_USE_AES_IF_THE_ALGORITHM_NEEDS_A_BLOCK_CIPHER(mac_alg)`, and at that point it's easier to just make HMAC and CMAC separate macros, with HMAC parametrized on the hash but CMAC just hard-coded to the one block cipher people use in practice.

**Comment by silabs-Kusumit** [comment]
I will change the `test_suite_psa_crypto_op_fail.generated` test data for the `PSA mac CMAC: !CMAC with AES` to also include `!PSA_WANT_PBKDF2_AES_CMAC_PRF_128`.  

The `config_psa.c` definition will look something like this:
```
#if defined(PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128)
#if !defined(MBEDTLS_PSA_ACCEL_ALG_PBKDF2_AES_CMAC_PRF_128)
#define MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_AES_CMAC_PRF_128 1
#if !defined(MBEDTLS_PSA_ACCEL_ALG_CMAC) || \
    defined(PSA_HAVE_SOFT_KEY_TYPE_AES)
#define MBEDTLS_PSA_BUILTIN_ALG_CMAC 1
#define MBEDTLS_CMAC_C
#endif /* !MBEDTLS_PSA_ACCEL_ALG_CMAC || PSA_HAVE_SOFT_KEY_TYPE_AES*/
#endif /* !MBEDTLS_PSA_ACCEL_ALG_PBKDF2_AES_CMAC_PRF_128 */
#endif /* PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128 */
```

**Comment by gilles-peskine-arm** [comment]
> I will change the test_suite_psa_crypto_op_fail.generated test data for the PSA mac CMAC: !CMAC with AES to also include !PSA_WANT_PBKDF2_AES_CMAC_PRF_128.

I don't think that should be needed. If our implementation delivers AES-CMAC, we should enable `PSA_WANT_ALG_CMAC` and `PSA_KEY_TYPE_AES`. We're already doing something similar for `PSA_WANT_KEY_TYPE_DERIVE` and friends, which are effectively always enabled, so we force them on at the bottom of `config_psa.h`.

**Comment by gilles-peskine-arm** [comment]
There's a known issue with the GitHub reporting at the moment. https://ci.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-head/detail/PR-7818-head/5/pipeline/ passed on d768ddf0d4eedfcaa64f6ca173733fad08af78cd.

**Comment by gilles-peskine-arm** [comment]
There's a conflict in the test data. Can you please rebase to resolve it?

**Comment by silabs-Kusumit** [comment]
Rebased to resolve conflicts

**Comment by gilles-peskine-arm** [comment]
The development branch was broken yesterday. We just merged a [fix](https://github.com/Mbed-TLS/mbedtls/pull/7870). Can you please rebase again? Sorry.

**Comment by silabs-Kusumit** [comment]
No problem

File: include/psa/crypto_builtin_key_derivation.h [hunk 1]
```diff
@@ -105,7 +105,8 @@ typedef struct psa_tls12_prf_key_derivation_s {
 } psa_tls12_prf_key_derivation_t;
 #endif /* MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF) ||
         * MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS */
-#if defined(MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_HMAC)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_HMAC) || \
```
Comment (tgonzalezorlandoarm) [inline include/psa/crypto_builtin_key_derivation.h:108]: What I'm wondering in general now is if PBKDF2 deverves a constant by itself so that we don't have to put a concatenation of 'defined' conditions for stuff that applies to all the PBKDF2 related algorithms. This would also apply, for example, in `include/psa/crypto_driver_contexts_key_derivation.h`.
@gilles-peskine-arm does what I'm saying make sense?

**Review by tgonzalezorlandoarm — COMMENTED** [review]
Initial General comments/questions.
I still need to review the tests (test vectors, testing logic, etc), which will also imply re-reviewing some `psa_crypto.c` functionality.

File: library/psa_crypto.c [hunk 2]
```diff
@@ -6159,10 +6180,14 @@ static psa_status_t psa_key_derivation_setup_kdf(
...
     /* All currently supported key derivation algorithms (apart from
-     * ecjpake to pms) are based on a hash algorithm. */
+     * ecjpake to pms and pbkdf2_aes_cmac_128) are based on a hash algorithm. */
     psa_algorithm_t hash_alg = PSA_ALG_HKDF_GET_HASH(kdf_alg);
     size_t hash_size = PSA_HASH_LENGTH(hash_alg);
-    if (kdf_alg != PSA_ALG_TLS12_ECJPAKE_TO_PMS) {
+    if (kdf_alg == PSA_ALG_TLS12_ECJPAKE_TO_PMS) {
+        hash_size = PSA_HASH_LENGTH(PSA_ALG_SHA_256);
+    } else if (kdf_alg == PSA_ALG_PBKDF2_AES_CMAC_PRF_128) {
+        hash_size = PSA_AES_CMAC_PRF_128_OUTPUT_SIZE;
```
[the remark this answers is not in the document]
Comment (silabs-Kusumit) [reply library/psa_crypto.c:6189]: Reworking this will be covered in the issue #7711

File: library/psa_crypto.c [hunk 3]
```diff
@@ -6159,10 +6180,14 @@ static psa_status_t psa_key_derivation_setup_kdf(
...
+    } else {
```
[the remark this answers is not in the document]
Comment (silabs-Kusumit) [reply library/psa_crypto.c:6190]: Reworking this will be covered in the issue https://github.com/Mbed-TLS/mbedtls/issues/7711

File: library/psa_crypto.c [hunk 4]
```diff
@@ -6159,10 +6180,14 @@ static psa_status_t psa_key_derivation_setup_kdf(
     }
 
     /* All currently supported key derivation algorithms (apart from
-     * ecjpake to pms) are based on a hash algorithm. */
+     * ecjpake to pms and pbkdf2_aes_cmac_128) are based on a hash algorithm. */
```
[the remark this answers is not in the document]
Comment (silabs-Kusumit) [reply library/psa_crypto.c:6183]: Reworking this will be covered in the issue https://github.com/Mbed-TLS/mbedtls/issues/7711

**Comment by tgonzalezorlandoarm** [comment]
In `crypto_config_profile_medium.h` I see the following comment:
```
/* PBKDF2-HMAC is not yet supported via the PSA API in Mbed TLS.
 * Note: when adding support, also adjust include/mbedtls/config_psa.h */
//#define PSA_WANT_ALG_PBKDF2_HMAC                1
```
Is this comment still applicable?
Should we make a similar comment in this PR as well?

**Comment by tgonzalezorlandoarm** [comment]
I'm failing to execute the tests:

`./test_suite_psa_crypto`

```
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, 16  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, 15  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, 0+  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 1 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 2 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 3 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 4 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 5 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 6 ........ ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, sa  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, pa  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, pa  ----
PSA key derivation: PBKDF2-AES-CMAC-PRF-128, OpenThread vector, pa  ----
PSA key derivation: ECJPAKE to PMS, no input ...................... PASS
```

This happens as well with HMAC:

```
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, 64+0 ........ ----
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, 54+10 ....... ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 20+0 .......... ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 0+20 .......... ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 1+19 .......... ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 10+10 ......... ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #2 ................ ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #3 ................ ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #5 ................ ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #6 ................ ----
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, salt in two   ----
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, password as   ----
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, password as   ----
PSA key derivation: PBKDF2-HMAC(SHA-256), RFC7914 #1, password as   ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, salt before co  ----
PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 20+1 (over cap  ----
```

What do I need to do to execute them?

**Comment by silabs-Kusumit** [comment]
You need to enable the `MBEDTLS_PSA_CRYPTO_CONFIG` in include/mbedtls/mbedtls_config.h#L2004

File: include/psa/crypto_builtin_key_derivation.h (hunk 1 above)
Comment (gilles-peskine-arm) [reply include/psa/crypto_builtin_key_derivation.h:108]: Yes, that makes sense, like we have `PSA_HAVE_SOFT_BLOCK_MODE` for generic block cipher handling code, we could have a `PSA_HAVE_SOFT_PBKDF2` for generic PBKDF2 code.

**Comment by gilles-peskine-arm** [comment]
> In crypto_config_profile_medium.h I see the following comment: (…)

That's copied from `include/psa/crypto_config.h`. `crypto_config_profile_medium.h` is a copy of a file from [TF-M](https://www.trustedfirmware.org/projects/tf-m/) and I think we shouldn't change it here: we should only change it when the file changes upstream (unless we need to adapt to a breaking change, but that's not the case here).

**Comment by gilles-peskine-arm** [comment]
> What do I need to do to execute them?

Normally, we'd include a new algorithm in the default configuration. However, because PBKDF2 support is still work in progress, we haven't enabled it yet. It should be enabled in the full config (i.e. by `scripts/config.py full`), because this is supposed to turn on even experimental features (it's supposed to activate everything that can be tested, except when there are contradictory options or platform-specific options). However, currently, `config.py full` doesn't enable PSA-only mechanisms,. But this is a defect that's overdue for a fix, so I'm adding it to the PBKDF2 epic. PR: https://github.com/Mbed-TLS/mbedtls/pull/7961

**Comment by gilles-peskine-arm** [comment]
Once we finish PBKDF2 (including the verify API), we should enable it in the default configuration and also announce it in a changelog entry. https://github.com/Mbed-TLS/mbedtls/issues/7962

**Comment by tgonzalezorlandoarm** [comment]
Thanks both. I was able to run the tests successfully and to understand why I was missing what I was missing. I added myself as a reviewer for #7962 for when it's ready to be reviewed.

**Comment by tgonzalezorlandoarm** [comment]
- It seems to me like we are not using any of [these test vectors](https://datatracker.ietf.org/doc/html/rfc4615#section-4). Maybe we should?
- It is not immediately obvious to me where Open Thread Vectors come from, could you point them to me and also does this deserve a comment somewhere?

File: tests/suites/test_suite_psa_crypto.data [hunk 5]
```diff
@@ -6325,6 +6401,59 @@ PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 20+1 (over capacity)
...
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"1b72f6419173a06e27777606a315876ec71227de":"":0:1:0
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 2
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"02":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"160597e28021fb3dd9cf088b007b688360fed438":"":0:1:0
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 3
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"1000":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"38ba9795fe87e47d519eacb77e82e35daa795870":"":0:1:0
```
Comment (tgonzalezorlandoarm) [inline tests/suites/test_suite_psa_crypto.data:6427]: I'm trying to recreate the test vector as per your instructions in the commit message.

I created a script (code below), and when I call `python3 myscript.py 70617373776f7264 73616c74 1000 20` I get a different result: `0141d00413c64e46440027166492ffabddc5b6b8` ( I was expecting `38ba9795fe87e47d519eacb77e82e35daa795870` )

The code for the script is:
```
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Hash import CMAC
from Crypto.Cipher import AES
import sys


def main():
    #check args
    if len(sys.argv) != 5:
        print("Invalid number of arguments. Expected: <password> <salt> <input_cost> <derived_key_len>")
        return

    password    = bytes.fromhex(sys.argv[1])
    salt        = bytes.fromhex(sys.argv[2])
    iterations  = int(sys.argv[3])
    dklen       = int(sys.argv[4])

    # If password is not 16 bytes then we need to use CMAC to derive the password
    if len(password) != 16:
        zeros     = bytes.fromhex("00000000000000000000000000000000")
        cobj_pass = CMAC.new(zeros, msg=password, ciphermod=AES, mac_len=16)
        passwd    = bytes.fromhex(cobj_pass.hexdigest())
    else:
        passwd = password

    cmac_prf = lambda p,s: CMAC.new(p, s, ciphermod=AES, mac_len=16).digest()

    actual_output = PBKDF2(passwd, salt=salt, dkLen=dklen, count=iterations, prf=cmac_prf)

    print('password         : ' + password.hex())
    print('salt             : ' + salt.hex())
    print('input cost       : ' + str(iterations))
    print('derived key len  : ' + str(dklen))
    print('output           : ' + actual_output.hex())

if __name__ == "__main__":
    main()
```

Please, could you help me understand what should be done differently?
Thanks

Comment (silabs-Kusumit) [reply tests/suites/test_suite_psa_crypto.data:6427]: The input cost in the test data is given as the hex value. 
The input cost to the test script is given as decimal integer. 

You need to pass `4096` instead of `1000`

**Comment by tgonzalezorlandoarm** [comment]
Okay, I finished with my review. Once the comments are addressed then this would look good to me :)

File: tests/suites/test_suite_psa_crypto.data [hunk 6]
```diff
@@ -6325,6 +6401,59 @@ PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 20+1 (over capacity)
...
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 5
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"1000":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"7361006c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"7061737300776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":16:"3d2828c5a437d781e7733ca353c40579":"":0:1:0
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test Vector 6
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"706173737764":PSA_SUCCESS:0:"":PSA_SUCCESS:"":64:"28e288c6345bb5ecf7ca70274208a3ba0f1148b5868537d5e09d3ee6813b1f524d9ecbf864eb814a46cda50ad5ec4c0dc03578c6c5fb4a3f9880deb5cab537e4":"":0:1:0
+
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:6440]: Missing: a test case that checks the expected output where the password is empty (potential edge case), and one where the password is exactly 16 bytes (known different behavior).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Looks mostly good to me

**Comment by silabs-Kusumit** [comment]
> * It seems to me like we are not using any of [these test vectors](https://datatracker.ietf.org/doc/html/rfc4615#section-4). Maybe we should?

We cannot use those test vectors as is. In PBKDF2 we append the block number to the salt and the block number starts from 1. Therefore even if we set the `password = key` and `salt = message`, PBKDF2 will process `message || 1`

**Review by tgonzalezorlandoarm — CHANGES_REQUESTED** [review]
looks good to me except for some minor comments. Also, I think that there is not clarification as to where the OpenThread vectors come from?

**Comment by daverodgman** [comment]
How does this relate to #806 ? It looks like that provides the same functionality, but via legacy interface rather than PSA? If that's the case, I suggest we close #806 when this PR is merged.

**Comment by gilles-peskine-arm** [comment]
@daverodgman Indeed: #806 is the same feature, same use case, different API.

**Comment by tgonzalezorlandoarm** [comment]
Just waiting for the OpenThread references and that's it

**Comment by silabs-Kusumit** [comment]
> * It is not immediately obvious to me where the OpenThread Vectors come from, could you point them to me and also does this deserve a comment somewhere?

The OpenThread vector is being used by us for internal testing. I tried to find it's source but was not able to trace it. 
Removing that from the PR now.

**Review by tgonzalezorlandoarm — DISMISSED** [review]

File: tests/suites/test_suite_psa_crypto.data [hunk 7]
```diff
@@ -6325,6 +6401,64 @@ PSA key derivation: PBKDF2-HMAC(SHA-1), RFC6070 #1, 20+1 (over capacity)
...
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"1b72f6419173a06e27777606a315876ec71227de":"":0:1:1
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test vector 1, password as bytes
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"1b72f6419173a06e27777606a315876ec71227de":"":0:0:0
+
+PSA key derivation: PBKDF2-AES-CMAC-PRF-128, Test vector 1, password as bytes, derive key
+depends_on:PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_WANT_ALG_CMAC:PSA_WANT_KEY_TYPE_AES
+derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"":"":0:0:1
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:6460]: `derive_output` ignores the output strings when outputting keys, but [it shouldn't](https://github.com/Mbed-TLS/mbedtls/issues/8038). In preparation for fixing this bug (which is out of scope here), please put the expected output here. Using your python script:

```suggestion
derive_output:PSA_ALG_PBKDF2_AES_CMAC_PRF_128:PSA_KEY_DERIVATION_INPUT_COST:"01":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_SALT:"73616c74":PSA_SUCCESS:PSA_KEY_DERIVATION_INPUT_PASSWORD:"70617373776f7264":PSA_SUCCESS:0:"":PSA_SUCCESS:"":20:"1b72f6419173a06e27777606a315876ec71227de":"":0:0:1
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
A couple of minor issues

**Comment by gilles-peskine-arm** [comment]
Speaking of OpenThread and test vectors: OpenThread comes with a [test hardness](https://github.com/openthread/openthread/blob/main/tools/harness-thci/OpenThread.py) which has the line
```
tools/harness-thci/OpenThread.py:from GRLLibs.UtilityModules.Plugins.AES_CMAC import Thread_PBKDF2
```
But I can't find where this `GRLLibs` module comes from. (Neither Google nor `requirements.txt` leads anywhere.) Do you happen to know where to find it?

File: include/mbedtls/config_psa.h [hunk 8]
```diff
@@ -439,13 +439,21 @@ extern "C" {
 #define PSA_HAVE_SOFT_BLOCK_AEAD 1
 #endif
 
+#if defined(PSA_WANT_ALG_PBKDF2_AES_CMAC_PRF_128)
+#if !defined(MBEDTLS_PSA_ACCEL_ALG_PBKDF2_AES_CMAC_PRF_128)
+#define MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_AES_CMAC_PRF_128 1
+#define PSA_HAVE_SOFT_PBKDF2_CMAC
```
[the remark this answers is not in the document]
Comment (silabs-Kusumit) [reply include/mbedtls/config_psa.h:445]: I had mainly added `PSA_HAVE_SOFT_PBKDF2_CMAC` for adding it to the definitions of `AES` and `CMAC` as using the `MBEDTLS_PSA_BUILTIN_ALG_PBKDF2_AES_CMAC_PRF_128` for enabling those features did not feel right. `PSA_HAVE_SOFT_PBKDF2_CMAC` is only used in the `config_psa.h`.

File: tests/suites/test_suite_psa_crypto.data (hunk 7 above)
Comment (silabs-Kusumit) [reply tests/suites/test_suite_psa_crypto.data:6460]: I will add the correct output and also correct the test case for PBKDF2-HMAC

**Comment by silabs-Kusumit** [comment]
> Speaking of OpenThread and test vectors: OpenThread comes with a [test hardness](https://github.com/openthread/openthread/blob/main/tools/harness-thci/OpenThread.py) which has the line
> 
> ```
> tools/harness-thci/OpenThread.py:from GRLLibs.UtilityModules.Plugins.AES_CMAC import Thread_PBKDF2
> ```
> 
> But I can't find where this `GRLLibs` module comes from. (Neither Google nor `requirements.txt` leads anywhere.) Do you happen to know where to find it?

I asked our OpenThread team and their response was that OpenThread.py is the script that is used by the Thread test harness to interface to the OpenThread API.  It is also known as the THCI file.  The Thread test harness is written by Granite River Labs (GRL) so the libs referenced must be to utilities provided by the Thread test harness. They are available when you download and run the test harness.

**Review by tgonzalezorlandoarm — APPROVED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

File: include/mbedtls/config_psa.h [hunk 9]
```diff
@@ -515,12 +524,18 @@ extern "C" {
 
 #if defined(PSA_WANT_ALG_CMAC)
 #if !defined(MBEDTLS_PSA_ACCEL_ALG_CMAC) || \
-    defined(PSA_HAVE_SOFT_BLOCK_CIPHER)
+    defined(PSA_HAVE_SOFT_BLOCK_CIPHER) || \
+    defined(PSA_HAVE_SOFT_PBKDF2_CMAC)
```
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply include/mbedtls/config_psa.h:528]: I think you're right: the PBKDF2-AES-CMAC code should work if AES-CMAC is provided by a driver. I'm [not sure if CMAC drivers are working](https://github.com/Mbed-TLS/mbedtls/issues/8314) though...

Since you're working on https://github.com/Mbed-TLS/mbedtls/pull/8346, would you mind checking if this exclusion can be removed, i.e. does the accelerated-CMAC driver work without `|| defined(PSA_HAVE_SOFT_PBKDF2_CMAC)` here and for AES?
